### PR TITLE
Implemented DCGAN Tests

### PR DESCRIPTION
**Pull request by ShikharJ**
Currently this contains the commits from the GAN PR, which should be merged soon, and then this would be rebased. Just the parameters of the generator network need to be corrected for this to work, which would be done soon.

### CODE REVIEW

**Comment by zoq** [comment]
Sounds like a good plan to me, thanks!

**Comment by zoq** [comment]
Okay, the ``Gradient``function is calling:

https://github.com/mlpack/mlpack/blob/6e0df8755b3c9ce3f1dccf8adda9d9747dd2874f/src/mlpack/methods/ann/ffn_impl.hpp#L218-L248

from the ``FFN`` class, which runs the Forward/Backward/Gradient function. I'll take a closer look into the issue tomorrow.

**Comment by zoq** [comment]
On a first look

```c++
generator.Add<TransposedConvolution<> >(4 * dNumKernels, 2 * dNumKernels, 4, 4, 2, 2, 1, 1,
      3, 3);
```

returns the wrong size in the backward pass.

**Comment by ShikharJ** [comment]
@zoq Silly me, I must have made a calculation mistake while setting the parameters. Sorry for bothering you on a weekend :)

**Comment by ShikharJ** [comment]
@zoq For the `10,000` image and `20 epoch` run on `MNIST`, seems like `DCGAN` is a bit slower to converge, but atleast the results are nice. Probably we should spend more time looking for hyper-parameters:
mnist_dcgan_20_epoch

**Comment by zoq** [comment]
Promising results, I think the main factor we are trying to do right now is to accelerate the training process right?

**Comment by ShikharJ** [comment]
@zoq Can you also help me in making the Travis build passable?

File: src/mlpack/tests/dcgan_test.cpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::math;
+using namespace mlpack::optimization;
+using namespace mlpack::regression;
+using namespace std::placeholders;
+
+// BOOST_AUTO_TEST_SUITE(DCGANNetworkTest);
```
Comment (rcurtin) [inline src/mlpack/tests/dcgan_test.cpp]: The CMake/CTest configuration will try to run the `DCGANNetworkTest` when this file is added regardless, so you could either comment the file out from `CMakeLists.txt`, uncomment this line (and maybe at least one test will need to be uncommented), or uncomment the whole thing.  I think that will fix the current Travis issue.

Comment (ShikharJ) [reply src/mlpack/tests/dcgan_test.cpp]: I commented out the file from CMakeLists.txt, it's better than letting the tests run, and potentially timeout the build.

File: src/mlpack/methods/ann/layer/leaky_relu.hpp [hunk 2]
```diff
@@ -46,7 +46,7 @@ class LeakyReLU
  public:
   /**
    * Create the LeakyReLU object using the specified parameters.
-   * The non zero gradient can be adjusted by specifying tha parameter
+   * The non zero gradient can be adjusted by specifying the parameter
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/leaky_relu.hpp]: Nice catch!

**Comment by rcurtin** [comment]
Should we uncomment the tests before merge?  Maybe you have already had a discussion about this and plan to implement them later or something, so it's possible I am just not aware of the plan. :)

File: src/mlpack/tests/dcgan_test.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  double stepSize = 0.0003;
+  double eps = 1e-8;
+  size_t numEpoches = 1;
+  double tolerance = 1e-5;
+  int datasetMaxCols = 40;
+  bool shuffle = true;
+  double multiplier = 10;
+  std::string output_dataset = "output_dcgan_mnist.csv";
+
+  std::cout << "output_dataset = '" << output_dataset << "'" << std::endl;
```
Comment (zoq) [inline src/mlpack/tests/dcgan_test.cpp]: hm, either we remove the information or use Log::Info/Log::Debug.

File: src/mlpack/tests/dcgan_test.cpp [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+  size_t generatorUpdateStep = 1;
+  size_t numSamples = 10;
+  double stepSize = 0.0003;
+  double eps = 1e-8;
+  size_t numEpoches = 1;
+  double tolerance = 1e-5;
+  int datasetMaxCols = 40;
+  bool shuffle = true;
+  double multiplier = 10;
+  std::string output_dataset = "output_dcgan_mnist.csv";
```
Comment (zoq) [inline src/mlpack/tests/dcgan_test.cpp]: We should use camel casing for all names.

File: src/mlpack/tests/dcgan_test.cpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    samples = trainData.col(math::RandInt(0, trainData.n_cols));
+    samples.reshape(dim, dim);
+    samples = samples.t();
+
+    generatedData.submat(dim,
+        i * dim, 2 * dim - 1, i * dim + dim - 1) = samples;
+  }
+
+  std::cout << "Saving output to " << output_dataset << "..." << std::endl;
+  generatedData.save(output_dataset, arma::csv_ascii);
```
Comment (zoq) [inline src/mlpack/tests/dcgan_test.cpp]: In an automated test case, I think we don't need to save the output, but we can comment the lines, and add a comment; I guess someone might find this helpful. Also, I'd like to note that we just test if the the code builds and runs, since we do not take a look at the output.

Comment (zoq) [reply src/mlpack/tests/dcgan_test.cpp]: If we just use this one to see if the code builds, perhaps we can decrease the number of training iterations?

Comment (ShikharJ) [reply src/mlpack/tests/dcgan_test.cpp]: The number of iterations is anyways down to the minimum (`1`).

**Comment by zoq** [comment]
```c++
size_t numIterations = trainData.n_cols * numEpoches;
numIterations /= batchSize;
```

isn't ``numIterations > 0`` since ``trainData.n_cols = datasetMaxCols = 40``?

**Comment by ShikharJ** [comment]
@zoq Ah, sorry, I got confused between number of epochs and number of iterations. The reason why I set it at 40 was because I could still get decent outputs of 4s within a single epoch of training (probably because the first 40 images must be images of 4s only). Hence if someone wants to save the generated output and visualize, he can easily do so. Also, it hardly affects the `Travis` runtime.

**Comment by zoq** [comment]
I see, do you think we could reduce the number of iterations, just to slightly reduce the runtime?

**Comment by ShikharJ** [comment]
@zoq This is good to go. The modifications made here would be helpful in implementing `WGAN`.
